### PR TITLE
JS: proper support for `this` inside a JSX-name

### PR DESCRIPTION
**Pull request by erik-krogh**
Previously `this` in something like the below would be parsed as an `Identifier` with the name `"this"`.   
With this change, we instead parse it as a `this`-expression. 

```JavaScript
class C extends SomeComponent {
  render() {
    return <this.MyComponent>foo</this.MyComponent>
  }
}
```

The expected output of the Trap test is updated because one of the tests contained a `<this.props.icon/>` element.

### CODE REVIEW

**Comment by esbena** [comment]
Good catch (or do we have an issue somewhere that mentions this?)!

This LGTM. Thanks for making the `ReactName.qll` tests for this change.

We just need to bump the extractor version string in Main.java.
(We could make a db upgrade script that updates the leading `this` identifiers to `this` expressions instead, would that be a trivial addition?)

**Comment by erik-krogh** [comment]
> Good catch (or do we have an issue somewhere that mentions this?)!

The latter.

> We just need to bump the extractor version string in Main.java.
> (We could make a db upgrade script that updates the leading `this` identifiers to `this` expressions instead, would that be a trivial addition?)

Possibly.. I don't think I've written such an upgrade before. But it doesn't look too difficult.

But doesn't that require that the database scheme has a new SHA? This PR required no change to `semmlecode.javascript.dbscheme`...

**Comment by esbena** [comment]
> But doesn't that require that the database scheme has a new SHA? (This PR required no change to semmlecode.javascript.dbscheme...)

Right. @asgerf what is the usual trick here, a dummy change?

File: javascript/extractor/src/com/semmle/js/ast/jsx/JSXThisExpr.java [hunk 1]
```diff
@@ -0,0 +1,21 @@
+package com.semmle.js.ast.jsx;
+
+import com.semmle.js.ast.ThisExpression;
+import com.semmle.js.ast.SourceLocation;
+import com.semmle.js.ast.Visitor;
+
+public class JSXThisExpr extends ThisExpression implements IJSXName {
```
Comment (asgerf) [inline javascript/extractor/src/com/semmle/js/ast/jsx/JSXThisExpr.java:7]: Is this class just here to create a `ThisExpression` that's also a `IJSXName`? If so, just make `ThisExpression` implement the `IJSXName` interface instead. That's what we do for `ITypeExpression`.

**Review by asgerf — COMMENTED** [review]
> (We could make a db upgrade script that updates the leading this identifiers to this expressions instead, would that be a trivial addition?)

Not worth doing IMO. We usually only do this if there are QL changes that are incompatible with an older database (such as changing the indexing of child nodes).

File: javascript/extractor/src/com/semmle/js/ast/jsx/JSXThisExpr.java (hunk 1 above)
Comment (erik-krogh) [reply javascript/extractor/src/com/semmle/js/ast/jsx/JSXThisExpr.java:7]: I did it like this to be consistent with the other classes that `implements IJSXName`, like: `public class JSXIdentifier extends Identifier implements IJSXName`.

**Review by asgerf — APPROVED** [review]

**Pull request merged**